### PR TITLE
[SofaKernel] Port Quat from Defaulttype to Sofa.Type

### PR DESCRIPTION
**Pull request by fredroy**
... and DualQuat.

## Description:
Move all Quaternion-related classes (and DualQuat) to the new Sofa.Type module.
Like #1734  There is a compatibility layer for transition, so it should be as least breaking as possible.
Remove helper::Quater/defaulttype::Quat thingy, really confusing (name "Quater" does not make any sense)

- sofa::helper::Quater<T> becomes sofa::type::Quat<T>
- sofa::defaulttype::Quat (typedef) becomes ... ?
- sofa::helper::DualQuat becomes sofa::type::DualQuat

## Breaking
- Remove a constructor with RigidType for DualQuat.
As DualQuat is only used in Flexible, not a big deal....

## Talking
 - Previous point: DualQuat is only used in Flexible, should it be simply moved there ?
 - fact that Quat is now a template (was a typedef without template before) could be confusing ?

## Note
Forgot wrapping redirection from MatSym, the fact that there was no error before means that it is not used at all...

Mandatory to mention: #1521

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 1]
```diff
@@ -177,22 +173,22 @@ class SOFA_HELPER_API Quater
...
 
 
-    /*! Returns the slerp interpolation of Quaternions \p a and \p b, at time \p t.
+    /*! Returns the slerp interpolation of Quatnions \p a and \p b, at time \p t.
 
      \p t should range in [0,1]. Result is \p a when \p t=0 and \p b when \p t=1.
 
      When \p allowFlip is \c true (default) the slerp interpolation will always use the "shortest path"
-     between the Quaternions' orientations, by "flipping" the source Quaternion if needed (see
+     between the Quatnions' orientations, by "flipping" the source Quatnion if needed (see
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]: Auto-replace issue :eyes:

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 2]
```diff
@@ -204,35 +200,35 @@ class SOFA_HELPER_API Quater
...
             Real nor = 1/phi;
             Real s = Real(sin(phi/2));
-            return Quater( a[0]*s*nor, a[1]*s*nor,a[2]*s*nor, Real(cos(phi/2)));
+            return Quat( a[0]*s*nor, a[1]*s*nor,a[2]*s*nor, Real(cos(phi/2)));
         }
     }
 
     /// Create a quaternion from Euler angles
-    /// Thanks to https://github.com/mrdoob/three.js/blob/dev/src/math/Quaternion.js#L199
+    /// Thanks to https://github.com/mrdoob/three.js/blob/dev/src/math/Quatnion.js#L199
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]: Auto-replace issue :eyes:

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 3]
```diff
@@ -570,30 +567,30 @@ defaulttype::Vec<3,Real> Quater<Real>::quatToRotationVector() const
...
 }
 
-/*! Returns the slerp interpolation of Quaternions \p a and \p b, at time \p t.
+/*! Returns the slerp interpolation of Quatnions \p a and \p b, at time \p t.
 
  \p t should range in [0,1]. Result is \p a when \p t=0 and \p b when \p t=1.
 
  When \p allowFlip is \c true (default) the slerp interpolation will always use the "shortest path"
- between the Quaternions' orientations, by "flipping" the source Quaternion if needed (see
+ between the Quatnions' orientations, by "flipping" the source Quatnion if needed (see
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]: Auto-replace issue :eyes:

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 4]
```diff
@@ -686,7 +683,7 @@ Quater<Real> Quater<Real>::slerp2(Quater<Real> &q1, Real t)
     }
     double ratioA = sin((1 - t) * halfTheta) / sinHalfTheta;
     double ratioB = sin(t * halfTheta) / sinHalfTheta;
-    //calculate Quaternion.
+    //calculate Quatnion.
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:686]: Auto-replace issue :eyes:

File: SofaKernel/modules/Sofa.Type/compat/sofa/defaulttype/MatSym.h [hunk 5]
```diff
@@ -24,7 +24,7 @@
 
 #include <sofa/type/MatSym.h>
 
-SOFA_DEPRECATED_HEADER(v21.12, "sofa/type/MatSym.h")
+// SOFA_DEPRECATED_HEADER(v21.12, "sofa/type/MatSym.h")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/compat/sofa/defaulttype/MatSym.h]: Add a quick comment to explain why we keep that.

File: SofaKernel/modules/Sofa.Type/compat/sofa/defaulttype/Quat.h [hunk 6]
```diff
@@ -19,24 +19,22 @@
...
-#define SOFA_DEFAULTTYPE_QUAT_H
+#pragma once
 
-#include <sofa/helper/Quater.h>
-#include <sofa/defaulttype/typeinfo/TypeInfo_Quat.h>
+#include <sofa/type/Quat.h>
 
-namespace sofa
-{
+// SOFA_DEPRECATED_HEADER(v21.12, "sofa/type/Quat.h")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/compat/sofa/defaulttype/Quat.h]: Add a quick comment to explain why we keep that.

File: SofaKernel/modules/Sofa.Type/compat/sofa/helper/DualQuat.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+
+#include <sofa/type/DualQuat.h>
+
+// SOFA_DEPRECATED_HEADER(v21.12, "sofa/type/DualQuat.h")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/compat/sofa/helper/DualQuat.h]: Add a quick comment to explain why we keep that.

File: SofaKernel/modules/Sofa.Type/compat/sofa/helper/DualQuat.inl [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+
+#include <sofa/type/DualQuat.inl>
+
+// SOFA_DEPRECATED_HEADER(v21.12, "sofa/type/DualQuat.inl")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/compat/sofa/helper/DualQuat.inl]: Add a quick comment to explain why we keep that.

File: SofaKernel/modules/Sofa.Type/compat/sofa/helper/Quater.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+
+#include <sofa/type/Quat.h>
+
+//SOFA_DEPRECATED_HEADER(v21.12, "sofa/type/Quat.h")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/compat/sofa/helper/Quater.h]: Add a quick comment to explain why we keep that.

File: SofaKernel/modules/Sofa.Type/compat/sofa/helper/Quater.inl [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+
+#include <sofa/type/Quat.inl>
+
+// SOFA_DEPRECATED_HEADER(v21.12, "sofa/type/Quat.inl")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/compat/sofa/helper/Quater.inl]: Add a quick comment to explain why we keep that.

File: SofaKernel/modules/Sofa.Type/src/sofa/type/MatSym.h [hunk 11]
```diff
@@ -457,7 +457,7 @@ bool invertMatrix(MatSym<S,real>& dest, const MatSym<S,real>& from)
 
         if (pivot <= (real) MIN_DETERMINANT)
         {
-            msg_error("MatSym") << "invertMatrix (general case) finds too small determinant: " << pivot << " for matrix = " << from;
+            //msg_error("MatSym") << "invertMatrix (general case) finds too small determinant: " << pivot << " for matrix = " << from;
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/MatSym.h]: To be removed?

File: SofaKernel/modules/Sofa.Type/src/sofa/type/MatSym.h [hunk 12]
```diff
@@ -512,7 +512,7 @@ bool invertMatrix(MatSym<3,real>& dest, const MatSym<3,real>& from)
 
     if ( -(real) MIN_DETERMINANT<=det && det<=(real) MIN_DETERMINANT)
     {
-        msg_error("MatSym") << "invertMatrix (special case 3x3) finds too small determinant: " << det << " for matrix = " << from;
+        //msg_error("MatSym") << "invertMatrix (special case 3x3) finds too small determinant: " << det << " for matrix = " << from;
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/MatSym.h]: To be removed?

File: SofaKernel/modules/Sofa.Type/src/sofa/type/MatSym.h [hunk 13]
```diff
@@ -534,7 +534,7 @@ bool invertMatrix(MatSym<2,real>& dest, const MatSym<2,real>& from)
 
     if ( -(real) MIN_DETERMINANT<=det && det<=(real) MIN_DETERMINANT)
     {
-        msg_error("MatSym") << "invertMatrix (special case 2x2) finds too small determinant: " << det << " for matrix = " << from;
+        //msg_error("MatSym") << "invertMatrix (special case 2x2) finds too small determinant: " << det << " for matrix = " << from;
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/MatSym.h]: To be removed?

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 14]
```diff
@@ -88,7 +84,7 @@ class SOFA_HELPER_API Quater
         return this->_q;
     }
 
-    /// Returns true if norm of Quaternion is one, false otherwise.
+    /// Returns true if norm of Quatnion is one, false otherwise.
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]: Auto-replace issue :eyes:

**Review by guparan — CHANGES_REQUESTED** [review]

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

**Pull request merged**